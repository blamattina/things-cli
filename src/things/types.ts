export interface CheckListJson {
  type: 'checklist-item';
  title: string;
  completed?: boolean;
}

export interface TodoJson {
  type: 'to-do';
  attributes: {
    'checklist-items'?: CheckListJson[];
    deadline?: string;
    list?: string;
    tags?: string[];
    title: string;
    when?: string;
    notes?: string;
  };
}

export interface HeadingJson {
  type: 'heading';
  atttributes: {
    title: string;
  };
}

export interface ProjectJson {
  type: 'project';
  attributes: {
    title: string;
    notes?: string;
    items: Array<TodoJson | HeadingJson>;
  };
}
