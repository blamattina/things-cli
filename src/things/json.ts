export interface ChecklistJson {
  type: 'checklist-item';
  attributes: {
    title: string;
    completed?: boolean;
  };
}

export interface TodoJson {
  type: 'to-do';
  attributes: {
    'checklist-items'?: ChecklistJson[];
    deadline?: string;
    list?: string;
    tags?: string[];
    title: string;
    when?: string;
    notes?: string;
  };
}

export interface HeadingJson {
  type: 'heading';
  atttributes: {
    title: string;
  };
}

export interface ProjectJson {
  type: 'project';
  attributes: {
    title: string;
    notes?: string;
    items: Array<TodoJson | HeadingJson>;
  };
}
