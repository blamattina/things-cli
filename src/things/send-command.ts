import openurl from 'openurl';

const encodeParams = function (params: Record<string, any>): string {
  if (!Object.keys(params).length) return '';

  return Object.keys(params).reduce(function (
    encoded: string,
    key: string,
    index: number,
  ) {
    encoded += index === 0 ? '?' : '&';
    encoded += `${encodeURIComponent(key)}=${encodeURIComponent(params[key])}`;
    return encoded;
  },
  '');
};

export const sendCommand = function (
  command: string,
  params = {},
  openUrl = openurl.open as Function,
) {
  const encodedParams = encodeParams(params || {});
  const url = `things:///${command}${encodedParams}`;
  openUrl(url);
};
