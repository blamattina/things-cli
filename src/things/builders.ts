import { TodoJson, ChecklistJson } from './json';
import { filterObject } from '../util/filterObject';

const buildChecklistItem = function (title: string): ChecklistJson {
  return {
    type: 'checklist-item',
    attributes: {
      title,
    },
  };
};

export const buildTodo = function ({
  title,
  list,
  when,
  checklist = [],
}: {
  title: string;
  list?: string;
  when?: string;
  checklist?: Array<string>;
}): TodoJson {
  return {
    type: 'to-do',
    attributes: {
      title,
      ...filterObject({
        list,
        when,
        'checklist-items': checklist.map(buildChecklistItem),
      }),
    },
  };
};
