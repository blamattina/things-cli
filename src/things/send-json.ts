import openurl from 'openurl';
import { TodoJson, ProjectJson } from './json';

export const sendJson = function (
  data = [] as Array<ProjectJson | TodoJson>,
  openUrl = openurl.open as Function,
) {
  const url = `things:///json?data=${encodeURIComponent(JSON.stringify(data))}`;
  openUrl(url);
};
