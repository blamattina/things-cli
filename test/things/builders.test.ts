import { buildTodo } from '../../src/things/builders';

describe('things/builders', () => {
  describe('buildTodo', () => {
    it('builds a todo with a title', () => {
      const todoJson = buildTodo({ title: 'test todo' });

      expect(todoJson).toMatchSnapshot();
    });

    it('builds a todo with an optional list', () => {
      const todoJson = buildTodo({ title: 'test todo', list: 'test list' });

      expect(todoJson).toMatchSnapshot();
    });

    it('builds a todo with an optional when', () => {
      const todoJson = buildTodo({ title: 'test todo', when: 'today' });

      expect(todoJson).toMatchSnapshot();
    });
  });
});
