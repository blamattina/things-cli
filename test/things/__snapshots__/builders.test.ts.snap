// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`things/builders buildTodo builds a todo with a title 1`] = `
Object {
  "attributes": Object {
    "title": "test todo",
  },
  "type": "to-do",
}
`;

exports[`things/builders buildTodo builds a todo with an optional list 1`] = `
Object {
  "attributes": Object {
    "list": "test list",
    "title": "test todo",
  },
  "type": "to-do",
}
`;

exports[`things/builders buildTodo builds a todo with an optional when 1`] = `
Object {
  "attributes": Object {
    "title": "test todo",
    "when": "today",
  },
  "type": "to-do",
}
`;
